pfio/MultiGroupServer.F90

```diff
@@ -418,301 +419,352 @@ end subroutine receive_output_data
-
-            else ! command /=1, notify the worker to quit and finalize
-
-               no_job = -1
-               do i = 1, this%nwriter -1
-                  if ( busy(i) == 0) then
-                     call MPI_send(no_job, 1, MPI_INTEGER, i, i, this%back_comm, ierr)
-                  else
-                     call MPI_recv( idle, 1, MPI_INTEGER, &
-                          i, stag, this%back_comm, &
-                          MPI_STAT, ierr)
-                     if (idle /= i ) stop ("idle should be i")
-                     call MPI_send(no_job, 1, MPI_INTEGER, i, i, this%back_comm, ierr)
-                  endif  
-               enddo  
-               exit
-            endif
-         enddo
-      else ! not root but writers 
-        do while (.true.)
+     _RETURN(_SUCCESS)
+
+   contains
+
+     subroutine start_back_captain(rc)
+       integer, optional, intent(out) :: rc
+       logical :: flag
+       integer :: nwriter_per_node
+       integer, allocatable :: idleRank(:,:)   ! idle processors  
+       integer, allocatable :: num_idlePEs(:) ! how many idle processors in each node of backend server 
+       integer :: i, no_job, local_rank, node_rank, nth_writer
+       integer :: terminate, idle_worker 
+
+       nwriter_per_node = this%nwriter/this%Node_Num
+       allocate(num_idlePEs(0:this%Node_Num-1))
+       allocate(idleRank(0:this%Node_Num-1, 0:nwriter_per_node-1), source=-1)
+
+        ! all local ranks are idle at the beginning
+       do local_rank = 1, this%nwriter-1
+          node_rank = this%node_ranks(local_rank)
+          nth_writer = mod(local_rank, nwriter_per_node)
+          idleRank(node_rank, nth_writer) = local_rank
+       enddo    
+
+       num_idlePEs = count(idleRank /=-1, dim = 2)
+
+       do while (.true.)
+         ! 1) captain node of back_comm is waiting command from front_comm
+         call MPI_recv( collection_id, 1, MPI_INTEGER, &
+              this%front_ranks(1), this%back_ranks(1), this%server_comm, &
+              MPI_STAT, ierr)
+
+         if (collection_id >= 1) then ! front_comm is asking for a writing node 
+            do while (.true.) ! try to retrieve idle writers
+              ! non block probe writers
+              do local_rank = 1, this%nwriter-1
+                flag = .false.
+                call MPI_Iprobe( local_rank, stag, this%back_comm, flag, MPI_STAT, ierr)             
+                if (flag) then
+                  call MPI_recv(idle_worker, 1, MPI_INTEGER, &
+                                local_rank, stag, this%back_comm, &
+                                MPI_STAT, ierr)
+                  _ASSERT(local_rank == idle_worker, "local_rank and idle_worker should match")
+                  node_rank = this%node_ranks(local_rank)
+                  num_idlePEs(node_rank) = num_idlePEs(node_rank) + 1
+                  nth_writer = mod(local_rank, nwriter_per_node)
```

Comment:
See previous comment I made. Should have made it here.

Reply:
I re-organized the codes before I saw your comment. Can you comment again and let me what is the problem?

Reply:
I think the fact that the first node has one processor less has been carefully taken care of all the time.